adserver/analyzer/tasks.py

```diff
@@ -47,36 +53,64 @@ def analyze_url(url, publisher_slug):
 
 
 @app.task
-def daily_visited_urls(days=2):
-    """Analyze URLs that were visited recently but have never been analyzed."""
-    dt_cutoff = timezone.now() - datetime.timedelta(days=days)
-
-    # Loop over each publisher and query the offers table for distinct URLs
-    for publisher in Publisher.objects.all():
-        for obj in (
-            Offer.objects.filter(publisher=publisher, date__gt=dt_cutoff)
-            .values("url")
-            .distinct()
-        ):
-            url = obj["url"]
-            if not url:
-                # Blank URLs can't be analyzed
-                continue
-
-            normalized_url = normalize_url(url)
-            if not AnalyzedUrl.objects.filter(
-                url=normalized_url, publisher_id=publisher.id
-            ).exists():
-                analyze_url.apply_async(args=[normalized_url, publisher.slug])
+def daily_visited_urls_aggregation(day=None):
+    """Aggregate URLs and update how many times they were visited in the last day."""
+    if day is None:
+        # Do the previous day's reports
+        day = timezone.now() - datetime.timedelta(days=1)
+
+    start_date, end_date = get_day(day)
+
+    # Query the offers table for distinct URLs
+    for obj in (
+        Offer.objects.using(settings.REPLICA_SLUG)
+        .filter(
+            date__gte=start_date,
+            date__lt=end_date,
+        )
+        .filter(viewed=True)
+        .filter(advertisement__flight__campaign__campaign_type=PAID)
+        .values("url", "publisher_id")
+        .annotate(visits=models.Count("url"))
+        .iterator()
+    ):
+        url = obj["url"]
+        publisher_id = obj["publisher_id"]
+        visits = obj["visits"]
+
+        if not url:
+            # Blank URLs can't be analyzed
+            continue
+
+        normalized_url = normalize_url(url)
+
+        url_obj, created = AnalyzedUrl.objects.get_or_create(
+            url=normalized_url,
+            publisher_id=publisher_id,
+            defaults={
+                "last_ad_served_date": timezone.now(),
+                "visits_since_last_analyzed": visits,
+            },
+        )
+
+        if not created:
+            # If the object wasn't created, add the visits and set the ad last served date
+            url_obj.update(
```

Comment:
Huh, I thought you could only call `update` on a queryset, not an object.

I wonder if we can this to make it a bit cleaner? https://docs.djangoproject.com/en/3.2/ref/models/querysets/#update-or-create

Reply:
Oooo, never used that before.

Reply:
Actually, since I'm using an `F()` query, I don't think I can use this method. I was looking forward to it.

Reply:
Gotcha. Still confused how we're calling `update` on a object instead of a queryset though, unless I'm reading the code wrong?

Reply:
Good catch. This is a problem. Adding a test for this and fixing it.